File: By-Python/leetcode/zigzag-conversion.py
class Solution(object):
    def convert(self, s, numRows):
        if numRows == 1 or len(s) <= numRows:
            return s
        
        groups = [""]*numRows
        idx, move = 0, 0
        
        for item in s:
            groups[idx] += item
            if idx == 0:
                move = 1
            elif idx == (numRows-1):
                move = -1
            idx += move
        
        return "".join(groups)

File: By-Swift/By-Swift/Leetcode/zigzag-conversion.swift
class Solution {
    func convert(_ s: String, _ numRows: Int) -> String {
        if numRows == 1 { return s }
        
        var groups = Array(repeating: "", count: numRows)
        var idx = 0
        var step = 1

        for l in Array(s) {
            groups[idx] += "\(l)"
            
            if idx+step == numRows {
                step = -1
            } else if idx+step < 0 {
                step = 1
            }
            idx += step
        }
        
        return groups.reduce(""){ $0 + $1 }
    }
}